pkg/query-service/app/inframetrics/hosts.go

```diff
@@ -0,0 +1,506 @@
+	groupKey := ""
+	for _, key := range groupBy {
+		groupKey += fmt.Sprintf("%s=%s,", key.Key, record.Meta[key.Key])
+	}
+	return groupKey
+}
+
+func (h *HostsRepo) getMetadataAttributes(ctx context.Context,
+	req model.HostListRequest, hostNameAttrKey string) (map[string]map[string]string, error) {
+	hostAttrs := map[string]map[string]string{}
+
+	hasHostName := false
+	for _, key := range req.GroupBy {
+		if key.Key == hostNameAttrKey {
+			hasHostName = true
+		}
+	}
+
+	if !hasHostName {
+		req.GroupBy = append(req.GroupBy, v3.AttributeKey{Key: hostNameAttrKey})
+	}
+
+	mq := v3.BuilderQuery{
+		AggregateAttribute: v3.AttributeKey{
+			Key:      "system_cpu_load_average_15m",
+			DataType: v3.AttributeKeyDataTypeFloat64,
+		},
+		Temporality: v3.Unspecified,
+		GroupBy:     req.GroupBy,
+	}
+	query, err := helpers.PrepareTimeseriesFilterQuery(req.Start, req.End, &mq)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(srikanthccv): remove this
+	query = strings.Replace(query, ".time_series_v4", ".distributed_time_series_v4", 1)
+
+	attrsListResponse, err := h.reader.GetListResultV3(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, row := range attrsListResponse {
+		stringData := map[string]string{}
+		for key, value := range row.Data {
+			if str, ok := value.(string); ok {
+				stringData[key] = str
+			} else if strPtr, ok := value.(*string); ok {
+				stringData[key] = *strPtr
+			}
+		}
+
+		hostName := stringData[hostNameAttrKey]
+		if _, ok := hostAttrs[hostName]; !ok {
+			hostAttrs[hostName] = map[string]string{}
+		}
+		for _, key := range req.GroupBy {
+			hostAttrs[hostName][key.Key] = stringData[key.Key]
+		}
+	}
+
+	return hostAttrs, nil
+}
+
+func (h *HostsRepo) getActiveHosts(ctx context.Context,
+	req model.HostListRequest, hostNameAttrKey string) (map[string]bool, error) {
+	activeStatus := map[string]bool{}
+	step := common.MinAllowedStepInterval(req.Start, req.End)
+
+	hasHostName := false
+	for _, key := range req.GroupBy {
+		if key.Key == hostNameAttrKey {
+			hasHostName = true
+		}
+	}
+
+	if !hasHostName {
+		req.GroupBy = append(req.GroupBy, v3.AttributeKey{Key: hostNameAttrKey})
+	}
+
+	params := v3.QueryRangeParamsV3{
+		Start: time.Now().Add(-time.Hour).UTC().UnixMilli(),
+		End:   time.Now().UTC().UnixMilli(),
+		Step:  step,
+		CompositeQuery: &v3.CompositeQuery{
+			BuilderQueries: map[string]*v3.BuilderQuery{
+				"A": {
+					QueryName:    "A",
+					StepInterval: step,
+					DataSource:   v3.DataSourceMetrics,
+					AggregateAttribute: v3.AttributeKey{
+						Key:      "system_cpu_load_average_15m",
+						DataType: v3.AttributeKeyDataTypeFloat64,
+					},
+					Temporality:      v3.Unspecified,
+					Filters:          req.Filters,
+					GroupBy:          req.GroupBy,
+					Expression:       "A",
+					TimeAggregation:  v3.TimeAggregationAvg,
+					SpaceAggregation: v3.SpaceAggregationAvg,
+					Disabled:         false,
```

Comment:
we should a limit for group by or is it fine like this ?

Reply:
The limit here won't do anything for metrics. The restriction on metric aggregation should be controlled using query limits, which is a separate exercise. But generally, we expect the number of hosts to be several thousand, so it should not be an issue. However, there are some metrics, such as disk, and network which will add noisy devices and cause issues. We address it by 1. query limits (separate PR https://github.com/SigNoz/signoz/pull/3956) 2. not collecting such datahttps://github.com/SigNoz/charts/pull/516